src/artemis/main.py

```diff
@@ -17,88 +25,131 @@
+            return StatusAndError(Status.FAILED, "Bluesky already stopping")
         else:
-            return StatusAndError(Status.IDLE)
+            self.current_status = StatusAndError(Status.ABORTING)
+            stopping_thread = threading.Thread(target=self.stopping_thread)
+            stopping_thread.start()
+            self.last_run_aborted = True
+            return StatusAndError(Status.ABORTING)
+
+    def shutdown(self):
+        """Stops the run engine and the loop waiting for messages."""
+        self.stop()
+        self.command_queue.put(Command(Actions.SHUTDOWN))
+
+    def wait_on_queue(self):
+        while True:
+            command = self.command_queue.get()
+            if command.action == Actions.START:
+                try:
+                    self.RE(get_plan(command.parameters))
+                    self.current_status = StatusAndError(Status.IDLE)
+                    self.last_run_aborted = False
+                except Exception as exception:
+                    if self.last_run_aborted:
+                        # Aborting will cause an exception here that we want to swallow
+                        self.last_run_aborted = False
+                    else:
+                        self.current_status = StatusAndError(
+                            Status.FAILED, str(exception)
+                        )
+            elif command.action == Actions.SHUTDOWN:
+                return
 
 
 class FastGridScan(Resource):
     def __init__(self, runner: BlueskyRunner) -> None:
         super().__init__()
         self.runner = runner
 
-    def _craft_return_message(self, status_and_message: StatusAndError):
-        return jsonify(
-            {
-                "status": status_and_message.status.value,
-                "message": status_and_message.error,
-            }
-        )
-
     def put(self, action):
         status_and_message = StatusAndError(Status.FAILED, f"{action} not understood")
         if action == Actions.START.value:
-            parameters = json.loads(request.data)
-            status_and_message = self.runner.start(parameters)
+            try:
+                parameters = FullParameters.from_json(request.data)
+                status_and_message = self.runner.start(parameters)
+            except JSONDecodeError as exception:
+                status_and_message = StatusAndError(Status.FAILED, str(exception))
         elif action == Actions.STOP.value:
             status_and_message = self.runner.stop()
-        return self._craft_return_message(status_and_message)
+        return status_and_message.to_dict()
 
     def get(self, action):
-        return self._craft_return_message(self.runner.get_status())
-
-
-def create_app(test_config=None) -> Tuple[Flask, BlueskyRunner]:
-    runner = BlueskyRunner()
+        if action == Actions.SHUTDOWN.value:
+            self.runner.shutdown()
+            shutdown_func = request.environ.get("werkzeug.server.shutdown")
```

Comment:
Not entirely clear what this is doing and why the shutdown action is in a GET request rather than PUT request like the other actions